src/storage/src/render/multi_worker_persist_sink.rs

```diff
@@ -0,0 +1,1123 @@
+        // through the map, so we cannot use the `mz_ore` wrapper either.
+        #[allow(clippy::disallowed_types)]
+        let mut in_flight_batches = std::collections::HashMap::<
+            (Antichain<Timestamp>, Antichain<Timestamp>),
+            Capability<Timestamp>,
+        >::new();
+
+        // TODO(aljoscha): We need to figure out what to do with error results from these calls.
+        let persist_client = persist_clients
+            .open(persist_location)
+            .await
+            .expect("could not open persist client");
+
+        let mut write = persist_client
+            .open_writer::<SourceData, (), Timestamp, Diff>(
+                shard_id,
+                &format!("compute::persist_sink::write_batches {}", collection_id),
+                Arc::new(target_relation_desc),
+                Arc::new(UnitSchema),
+            )
+            .await
+            .expect("could not open persist shard");
+
+        // The current input frontiers.
+        let mut batch_descriptions_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+        let mut desired_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+
+        // The frontiers of the inputs we have processed, used to avoid redoing work
+        let mut processed_desired_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+        let mut processed_descriptions_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+
+        // A "safe" choice for the lower of new batches we are creating.
+        let mut operator_batch_lower = Antichain::from_elem(TimelyTimestamp::minimum());
+
+        loop {
+            tokio::select! {
+                Some(event) = descriptions_input.next_mut() => {
+                    match event {
+                        Event::Data(cap, data) => {
+                            // Ingest new batch descriptions.
+                            for description in data.drain(..) {
+                                if collection_id.is_user() {
+                                    trace!(
+                                        "persist_sink {collection_id}/{shard_id}: \
+                                            write_batches: \
+                                            new_description: {:?}, \
+                                            desired_frontier: {:?}, \
+                                            batch_descriptions_frontier: {:?}",
+                                        description,
+                                        desired_frontier,
+                                        batch_descriptions_frontier,
+                                    );
+                                }
+                                let existing = in_flight_batches.insert(
+                                    description.clone(),
+                                    cap.delayed(description.0.first().unwrap()),
+                                );
+                                assert!(
```

Comment:
we should do this instead to avoid the unconditional clone of the description above:

```rust
if let Some(description) = existing {
    panic!(...);
}
```

Reply:
this became way more annoying because the description is the key, but the entry api worked